GitUpKit/Core/GCRepository+Bare.m

```diff
@@ -362,27 +366,67 @@ - (GCCommit*)createCommitFromTree:(git_tree*)tree
                             count:(NSUInteger)count
                            author:(const git_signature*)author
                           message:(NSString*)message
+                       shouldSign:(BOOL)shouldSign
                             error:(NSError**)error {
   GCCommit* commit = nil;
   git_signature* signature = NULL;
 
   git_oid oid;
+  
+  GCConfigOption* shouldSignOption = [self readConfigOptionForVariable:@"commit.gpgsign" error:nil];
   CALL_LIBGIT2_FUNCTION_GOTO(cleanup, git_signature_default, &signature, self.private);
-  CALL_LIBGIT2_FUNCTION_GOTO(cleanup, git_commit_create, &oid, self.private, NULL, author ? author : signature, signature, NULL, GCCleanedUpCommitMessage(message).bytes, tree, count, parents);
+  
+  git_buf commitBuffer = GIT_BUF_INIT_CONST("", 0);
+  const char *gpgSignature = NULL;
+  if ([shouldSignOption.value isEqualToString:@"true"]) {
+    GCConfigOption* signingKeyOption = [self readConfigOptionForVariable:@"user.signingkey" error:nil];
+    
+    CALL_LIBGIT2_FUNCTION_GOTO(cleanupBuffer, git_commit_create_buffer, &commitBuffer, self.private, author ? author : signature, signature, NULL, GCCleanedUpCommitMessage(message).bytes, tree, count, parents);
+    gpgSignature = [self gpgSig:commitBuffer.ptr keyId:signingKeyOption.value];
+  }
+  
+  if (gpgSignature != NULL) {
+    CALL_LIBGIT2_FUNCTION_GOTO(cleanupBuffer, git_commit_create_with_signature, &oid, self.private, commitBuffer.ptr, gpgSignature, NULL);
```

Comment:
@lucasderraugh Separate thread regarding [git-up/libgit2](https://github.com/git-up/libgit2) fork.
It looks like libgit2 doesn't implement 100% gpg support yet so I have to implement some additional methods in it. 
Is there any process for fork updates? Should it be just PR with functionality or I have to implement some tests in libgit2 as well? 

Generally, what was the motivation to link against fork instead of upstream? Would it be possible to rebase on top of upstream? Is there a chance that we may migrate to the upstream instead of the fork?

/cc @pks-t @swisspol

Reply:
I created a fork of libgit2 because a few patches were needed with some new APIs, bug fixes and change of behavior. Not all would be accepted in upstream and at the time maintainers could take weeks to respond. You can see the 15 or so custom commits here:
https://github.com/git-up/libgit2/commits/gitup

The README in the fork repo explains how to update to upstream.

Reply:
@swisspol – looks like we standstill with libgit2, maybe suggested override below can help? (i think libgit2 are bit resistant to add features, even signing – they usually say it's more to the clients)

but gosh, I'm missing that in GitUp so much :) hope resolution will be found. Also, please, lmk if we can support somehow?